lib/rpc/singleJsonRpcProvider.ts

```diff
@@ -0,0 +1,460 @@
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('getGasPrice')
+      })
+  }
+
+  override getLogs(filter: Filter): Promise<Array<Log>> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .getLogs(filter)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('getLogs')
+      })
+  }
+
+  override getNetwork(): Promise<Network> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .getNetwork()
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('getNetwork')
+      })
+  }
+
+  override getStorageAt(
+    addressOrName: string | Promise<string>,
+    position: BigNumberish | Promise<BigNumberish>,
+    blockTag?: BlockTag | Promise<BlockTag>
+  ): Promise<string> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .getStorageAt(addressOrName, position, blockTag)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('getStorageAt')
+      })
+  }
+
+  override getTransaction(transactionHash: string | Promise<string>): Promise<TransactionResponse> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .getTransaction(transactionHash)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('getTransaction')
+      })
+  }
+
+  override getTransactionCount(
+    addressOrName: string | Promise<string>,
+    blockTag?: BlockTag | Promise<BlockTag>
+  ): Promise<number> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
```

Comment:
i see there is a lot of duplication around 

```
    const startTime = Date.now()
    this.recordPerfBeforeCall(startTime)
```

Would it make more sense to make all these super calls encapsulated somehow so that we dont do this everytime? I think this might make this a lot more readable! We have certain examples on data-api-graphql

Reply:
practically, the only variable is the function call (`getTransactionCount` or `getTransactionReceipt` etc) so we can do something cool here